### PR TITLE
RUBY-2869 Do not use mutexes in finalizers

### PR DESCRIPTION
**Pull request by comandeo**
Finalizers are called from signal trap, so mutexes do not work there (see https://github.com/ruby/ruby/blob/master/thread_sync.c#L288).

### CODE REVIEW

**Review by Neilshweky — APPROVED** [review]

File: lib/mongo/cluster/reapers/cursor_reaper.rb [hunk 1]
```diff
@@ -56,12 +62,14 @@ def initialize(cluster)
...
-          end
-        end
+        msg = [
+          server.address.seed,
+          kill_spec.cursor_id,
+          kill_spec.coll_name,
+          kill_spec.db_name,
+          kill_spec.service_id
+        ].join(MSG_SEPARATOR)
+        @pipe_write.puts(msg)
```
Comment (p-mongo) [inline lib/mongo/cluster/reapers/cursor_reaper.rb:72]: This call can block if the pipe is full. My understanding is there needs to be  a thread that is constantly reading the pipe to ensure the writing side doesn't block indefinitely.

Comment (p-mongo) [reply lib/mongo/cluster/reapers/cursor_reaper.rb:72]: Have you considered https://ruby-doc.org/core-2.5.0/Queue.html ?

Comment (comandeo) [reply lib/mongo/cluster/reapers/cursor_reaper.rb:72]: I have, but `Queue` is thread safe, and according to its docs _"The [Queue](https://ruby-doc.org/core-2.5.0/Queue.html) class implements all the required locking semantics."_ So, we may encounter the same problem with queue. I'll do some testing, though. If it works, I would prefer queue over pipe here.

**Comment by comandeo** [comment]
> I think it is more correct to use server addresses (`Address` instances) rather than seeds. The seeds may be missing port numbers.

Good point, changed accordingly.

**Pull request merged**